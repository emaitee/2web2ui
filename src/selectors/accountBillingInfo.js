import _ from 'lodash';
import _fp from 'lodash/fp';

export function overviewProps(state) {
  const { subscription } = state.account;

  return {
    loading: !Object.keys(state.account).length || state.billing.plansLoading,
    hasBillingAccount: !!state.billing,
    plans: selectPublicPlans(state),
    currentPlan: selectCurrentPlan(state),
    subscription
  };
}

/**
 * Get public plans from state and sorts them by volume
 */
export function selectPublicPlans(state) {
  return state.billing.plans
   ? _.sortBy(state.billing.plans.filter((plan) => plan.status === 'public'), (plan) => plan.volume)
   : [];
}

/**
 * Gets current plan
 */
export function selectCurrentPlan(state) {
  return state.billing.plans && Object.keys(state.account).length
    ? _.find(state.billing.plans, { code: state.account.subscription.code })
    : {};
}

/**
 * Gets IP pools and formats for select options
 */
export function selectIpPools(state) {
  return _fp.flow(
    _fp.reject({ id: 'default' }),
    _fp.sortBy(({ name }) => name.toLowerCase()),
    _fp.map(({ name, id }) => ({ label: `${name} [${id}]`, value: id }))
  )(state.ipPools.list);
}
