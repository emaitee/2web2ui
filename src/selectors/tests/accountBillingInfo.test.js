import { overviewProps, selectIpPools } from '../accountBillingInfo';

describe('Selector: Get Account Overview Info', () => {

  let store;

  beforeEach(() => {
    store = {
      account: {
        subscription: {
          plan: 'test plan',
          code: 'test-plan-0817',
          self_serve: true
        },
      },
      billing: {
        plansLoading: false,
        plans: [
          {
            code: 'test-plan-0817',
            status: 'public',
            volume: 5000000
          },
        ]
      }
    };
  })

  it('should return the selected data for a customer with a billing account', () => {
    expect(overviewProps(store)).toMatchSnapshot();
  });
});

describe('IP Pool List Selector', () => {

  const store = {
    ipPools: {
      list: [
        { id: 'ip_id', name: 'IP Name', },
        { id: 'default', name: 'IP Name2', }
      ]
    }
  }

  it('should format IP Pools for select options', () => {
    expect(selectIpPools(store)).toMatchSnapshot();
  });

  it('should return empty array with no IP Pools', () => {
    store.ipPools = {};
    expect(selectIpPools(store)).toMatchSnapshot();
  });
});
