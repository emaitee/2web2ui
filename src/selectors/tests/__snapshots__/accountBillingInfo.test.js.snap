// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IP Pool List Selector should format IP Pools for select options 1`] = `
Array [
  Object {
    "label": "IP Name [ip_id]",
    "value": "ip_id",
  },
]
`;

exports[`IP Pool List Selector should return empty array with no IP Pools 1`] = `Array []`;

exports[`Selector: Get Account Overview Info should return the selected data for a customer with a billing account 1`] = `
Object {
  "currentPlan": Object {
    "code": "test-plan-0817",
    "status": "public",
    "volume": 5000000,
  },
  "hasBillingAccount": true,
  "loading": false,
  "plans": Array [
    Object {
      "code": "test-plan-0817",
      "status": "public",
      "volume": 5000000,
    },
  ],
  "subscription": Object {
    "code": "test-plan-0817",
    "plan": "test plan",
    "self_serve": true,
  },
}
`;
